src/lpython/parser/tokenizer.re

```diff
@@ -579,6 +593,59 @@ int Tokenizer::lex(Allocator &al, YYSTYPE &yylval, Location &loc, diag::Diagnost
     }
 }
 
+void Tokenizer::lex_match_or_case(Location &loc, unsigned char *cur,
+        bool &is_match_or_case_keyword) {
+    for (;;) {
+        unsigned char *tok = cur;
+        unsigned char *mar;
+        /*!re2c
+            re2c:define:YYCURSOR = cur;
+            re2c:define:YYMARKER = mar;
+            re2c:yyfill:enable = 0;
+            re2c:define:YYCTYPE = "unsigned char";
+
+            * {
+                token_loc(loc);
+                std::string t = std::string((char *)tok, cur - tok);
+                throw LFortran::parser_local::TokenizerError("Token '" + t
+                    + "' is not recognized in `match` statement", loc);
+            }
+
+            end {
+                token_loc(loc);
+                std::string t = std::string((char *)tok, cur - tok);
+                throw LFortran::parser_local::TokenizerError(
+                    "End of file not expected within `match` statement: '" + t
+                    + "'", loc);
+            }
+
+            "(" { record_paren(loc, '('); continue; }
+            "[" { record_paren(loc, '['); continue; }
+            "{" { record_paren(loc, '{'); continue; }
+            ")" { record_paren(loc, ')'); continue; }
+            "]" { record_paren(loc, ']'); continue; }
+            "}" { record_paren(loc, '}'); continue; }
+
+            ":" newline | ":" whitespace? comment {
+                if (parenlevel == 0) {
+                    is_match_or_case_keyword = true;
+                    return;
+                }
+                continue;
+            }
+
+            "\\" newline { continue; }
+
```

Comment:
I think one has to add a string matcher here, because somebody could have done `match (f("(")):`